### PR TITLE
Update Tensorflow 2.4.1

### PR DESCRIPTION
**Pull request by smuzaffar**
- Tensorflow: `2.4.1` only for python3
  -  Dropped `python2` support for `tensorflow` and `tensorboard`
  - `Tensorflow 2.4.1` only builds for py3. There are changes in TF python code which do not work/run with python2.  
  - use cms `GRPC` and `typing_extensions`
- GRPC: `1.35.0` which builds with `c++17`. This also needs `protobuf 3.12+`.
  - It builds fine with system `OpenSSL` but needs patches for `CMS OpenSSL`.
- Protobuf: `3.15.1`
  - Build with `c++17`
- OpenCV: `4.5.1`
  - For `ppc64le` build without eigen ( https://github.com/opencv/opencv/issues/19647 )
- Eigen: used commit `011e0db31d1bed8b7f73662be6d57d9f30fa457a`  from master branch

This still needs CMSSW TF code updates
FYI @riga

### CODE REVIEW

**Comment by smuzaffar** [comment]
please test with cms-sw/cmssw#32993

**Comment by smuzaffar** [comment]
please test for cc8_amd64_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
@fwyzard , we need to update Eigen to go with this TF update. TF is using 011e0db31d1bed8b7f73662be6d57d9f30fa457a eigen commit. I have created `cms/master/011e0db31d1bed8b7f73662be6d57d9f30fa457a` branch for https://github.com/cms-externals/eigen-git-mirror . I tried to get your fixes for cuda ( https://github.com/eigenteam/eigen-git-mirror/compare/d812f411c3f9...cms-externals:cms/master/d812f411c3f9 ) on top of `011e0db31d1bed8b7f73662be6d57d9f30fa457a` but there are too many conflicits. Can you please check if these are still needed and can you please provide PR to go on top of `cms/master/011e0db31d1bed8b7f73662be6d57d9f30fa457a`  branch?

**Comment by fwyzard** [comment]
I will be able to have a look only late next week (or later).

**Comment by smuzaffar** [comment]
please test with cms-sw/cmssw#32993

**Comment by smuzaffar** [comment]
please test with cms-sw/cmssw#32993

**Comment by smuzaffar** [comment]
test parameters:
- pull_request = cms-sw/cmssw#32993

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test

**Comment by riga** [comment]
@smuzaffar Nice!
I guess we can close https://github.com/cms-sw/cmsdist/pull/6545 when this one is merged.
I'll open a PR with required updates in the CMSSW interface early next week.

File: eigen.spec [hunk 1]
```diff
@@ -1,21 +1,14 @@
-### RPM external eigen d812f411c3f9
+### RPM external eigen 011e0db31d1bed8b7f73662be6d57d9f30fa457a
```
Comment (fwyzard) [inline eigen.spec:1]: @smuzaffar how did you choose [011e0db3](https://gitlab.com/libeigen/eigen/-/commit/011e0db31d1bed8b7f73662be6d57d9f30fa457a) ?

**Comment by fwyzard** [comment]
@smuzaffar can you ask the bot to build the `CMSSW_11_3_Patatrack_X` branch with these changes ?

File: eigen.spec (hunk 1 above)
Comment (fwyzard) [reply eigen.spec:1]: ah, sorry, I missed in your comment before that this comes from tensorflow

could we try also a newer version from the Eigen's master branch ?

Comment (smuzaffar) [reply eigen.spec:1]: we can try latest eigen (TF might fail if there are interface changes though)

**Comment by smuzaffar** [comment]
please test for CMSSW_11_3_Patatrack_X

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
@riga, `opencv` does not build for ppc64le with newer eigen. I have opened an issue https://github.com/opencv/opencv/issues/19647 here. So looks like this is a blocker or we disabled building opencv with eigen for ppc64le.

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test for CMSSW_11_3_Patatrack_X

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by fwyzard** [comment]
I can confirm that all the usual GPU workflows do work with the HEAD of Eigen.
However, I will need a bit more time to check if the performance is also the same, or if there is any degradation.

**Comment by smuzaffar** [comment]
@fwyzard , should I try thr head of eigen 3.3 branch or head of eigen master branch?

**Comment by fwyzard** [comment]
the master branch

**Comment by smuzaffar** [comment]
@fwyzard , tensorflow 2.4.1 does not build with latest eigen so for now I would suggest to stick to commit `011e0db31d1bed8b7f73662be6d57d9f30fa457a`

**Comment by fwyzard** [comment]
sigh

OK, I'll test if `011e0db31d1bed8b7f73662be6d57d9f30fa457a` works for the various GPU workflows

**Comment by fwyzard** [comment]
For the pixel part of the code, `011e0db31d1bed8b7f73662be6d57d9f30fa457a` builds fine.
Still need to test ECAL, HCAL and the performance.

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test for slc7_aarch64_gcc9

**Comment by smuzaffar** [comment]
please test for slc7_ppc64le_gcc9

**Comment by fwyzard** [comment]
OK, I was finally able to validate all GPU-related workflows with the new Eigen version, and didn't find any regression in the performance:

Eigen version | old | new
---|---|---
I/O only  |    347.2 ±   0.3 ev/s |   346.7 ±   0.8 ev/s
11634.501 |    327.6 ±   0.5 ev/s |   327.3 ±   0.5 ev/s
11634.502 |    344.9 ±   0.4 ev/s |   346.0 ±   0.3 ev/s
11634.505 |    311.2 ±   0.8 ev/s |   309.8 ±   1.0 ev/s
11634.506 |    316.9 ±   1.0 ev/s |   317.7 ±   0.4 ev/s
11634.511 |    322.0 ±   0.4 ev/s |   322.3 ±   0.4 ev/s
11634.512 |    345.7 ±   1.5 ev/s |   346.1 ±   0.8 ev/s
11634.521 |    325.3 ±   1.0 ev/s |   324.2 ±   0.1 ev/s
11634.522 |    346.9 ±   0.4 ev/s |   345.7 ±   0.8 ev/s

**Comment by smuzaffar** [comment]
+externals
all looks good for this PR to go in. @riga do you want to do any final performance tests before we merge it?

**Comment by smuzaffar** [comment]
lets get this in IBs

**Pull request merged**